tests/test_serializable_object.py

```diff
@@ -177,5 +136,131 @@ def test_cycle_detection(self):
+        with self.assertRaises(ValueError):
+            @otio.core.register_type
+            class DoubleReg(otio.core.SerializableObject):
+                _serializable_label = "Stuff.1"
+
+    def test_upgrade_versions(self):
+        """Test adding an upgrade functions for a type"""
+
+        @otio.core.register_type
+        class FakeThing(otio.core.SerializableObject):
+            _serializable_label = "NewStuff.4"
+            foo_two = otio.core.serializable_field("foo_2")
+
+        @otio.core.upgrade_function_for(FakeThing, 2)
+        def upgrade_one_to_two(_data_dict):
+            return {"foo_2": _data_dict["foo"]}
+
+        @otio.core.upgrade_function_for(FakeThing, 3)
+        def upgrade_one_to_two_three(_data_dict):
+            return {"foo_3": _data_dict["foo_2"]}
+
+        # not allowed to overwrite registered functions
+        with self.assertRaises(ValueError):
+            @otio.core.upgrade_function_for(FakeThing, 3)
+            def upgrade_one_to_two_three_again(_data_dict):
+                raise RuntimeError("shouldn't see this ever")
+
+        ft = otio.core.instance_from_schema("NewStuff", 1, {"foo": "bar"})
+        self.assertEqual(ft._dynamic_fields['foo_3'], "bar")
+
+        ft = otio.core.instance_from_schema("NewStuff", 3, {"foo_2": "bar"})
+        self.assertEqual(ft._dynamic_fields['foo_3'], "bar")
+
+        ft = otio.core.instance_from_schema("NewStuff", 4, {"foo_3": "bar"})
+        self.assertEqual(ft._dynamic_fields['foo_3'], "bar")
+
+    def test_upgrade_rename(self):
+        """test that upgrading system handles schema renames correctly"""
+
+        @otio.core.register_type
+        class FakeThingToRename(otio.core.SerializableObject):
+            _serializable_label = "ThingToRename.2"
+            my_field = otio.core.serializable_field("my_field", doc="example")
+
+        thing = otio.core.type_version_map()
+        self.assertTrue(thing)
+
+    def test_downgrade_version(self):
+        """ test a python defined downgrade function"""
+
+        @otio.core.register_type
+        class FakeThing(otio.core.SerializableObject):
+            _serializable_label = "FakeThingToDowngrade.2"
+            foo_two = otio.core.serializable_field("foo_2")
+
+        @otio.core.downgrade_function_for(FakeThing, 2)
+        def downgrade_2_to_1(_data_dict):
+            return {"foo": _data_dict["foo_2"]}
+
+        # not allowed to overwrite registered functions
+        with self.assertRaises(ValueError):
+            @otio.core.downgrade_function_for(FakeThing, 2)
+            def downgrade_2_to_1_again(_data_dict):
+                raise RuntimeError("shouldn't see this ever")
+
+        f = FakeThing()
+        f.foo_two = "a thing here"
+
+        # downgrade_target = otio.core.fetch_version_map(
+        #         "OTIO_CORE",
+        #         otio.__version__
+        # )
+
+        downgrade_target = {"FakeThingToDowngrade": 1}
+
+        result = eval(
```

Comment:
Eval is usually frown upon. Would a `json.loads` or even a `ast.literal_eval` do the trick?